### PR TITLE
Introduced p::d::CellDataTransfer.

### PR DESCRIPTION
**Pull request by marcfehling**
Recently, it turned out that we need a class that transfers cell by cell data across refinement and serialization (see #7515). This pull requests provides such a class for the `parallel::distributed::Triangulation`. We need another variant for the other cases as well.

In a follow-up pull request, this newly introduced class `p::d::CellDataTransfer` will replace `p::d::ActiveFEIndicesTransfer`, which has been introduced a couple of months ago, and we will hide the whole `active_fe_transfer` business in the `hp::DoFHandler` class, including serialization. Thus for now, we only provide a test for transfer across refinement.

Part of #3511.

### CODE REVIEW

File: include/deal.II/distributed/cell_data_transfer.h [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace distributed
+  {
+    /**
+     * Transfer data that is associated with each cell (like error indicators)
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: active cell

File: source/distributed/cell_data_transfer.inst.in [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+
+for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS)
+  {
+#if deal_II_dimension <= deal_II_space_dimension
+    // we need this specialization to transfer active fe indices
+    template class parallel::distributed::CellDataTransfer<
+      deal_II_dimension,
+      deal_II_space_dimension,
+      std::vector<unsigned int>>;
```
Comment (marcfehling) [inline source/distributed/cell_data_transfer.inst.in]: We introduced this specialization for `std::vector<unsigned int>>` to be able to transfer integers. We could also specialize for `dealii::Vector<unsigned int>`, but I think we will only use the integer transfer internally, so there is no need for it. What do you think?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+     * while refining and/or coarsening a distributed grid and handles the
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: grid -> triangulation

File: include/deal.II/distributed/cell_data_transfer.h [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+     * necessary communication.
+     *
+     * This class therefore does for cell-related information what
+     * parallel::distributed::SolutionTransfer does for the values of degrees of
+     * freedom defined on a parallel::distributed::Triangulation.
+     *
+     * This class has been designed to operate on miscellaneous data generated
+     * from current solutions that will be stored in non-distributed Vector
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: can we be more general here and support `std::vector<T>` where `T` is serializable? What if each cell contains `T=std::pair<double,double>` or an `T=std::set<unsigned int>`?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace distributed
+  {
+    /**
+     * Transfer data that is associated with each active cell (like error
+     * indicators) while refining and/or coarsening a distributed triangulation
+     * and handles the necessary communication.
```
Comment (tamiko) [inline include/deal.II/distributed/cell_data_transfer.h]:
```suggestion
     * and handle the necessary communication.
```

File: include/deal.II/distributed/cell_data_transfer.h [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+     *
+     * This class therefore does for cell-related information what
+     * parallel::distributed::SolutionTransfer does for the values of degrees of
+     * freedom defined on a parallel::distributed::Triangulation.
+     *
+     * This class has been designed to operate on any kind of datatype that is
+     * serializable. A non-distributed Vector container holding the cell by cell
+     * data has to be provided. Its size has to correspond to
+     * Triangulation::n_active_cells().
```
Comment (tamiko) [inline include/deal.II/distributed/cell_data_transfer.h]: When you say "holding the cell by cell data has to be provided" can you also quickly comment on what cells this refers to? (see my comment for `typename VectorType` below)

File: include/deal.II/distributed/cell_data_transfer.h [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+     *
+     * <h3>Use for serialization</h3>
+     *
+     * This class can be used to serialize and later deserialize a distributed
+     * mesh with attached data to separate files. If you use more than one
+     * parallel::distributed::CellDataTransfer object, they need to be
+     * serialized and deserialized in the same order.
+     *
+     * For serialization, the following code snippet saves not only the
+     * triangulation itself, but also the cell by cell data attached:
```
Comment (tamiko) [inline include/deal.II/distributed/cell_data_transfer.h]: maybe "cell-wise" instead of "cell by cell"?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+     * parallel::distributed::SolutionTransfer::deserialize(),
+     * need to be in the same order, respectively.
+     *
+     * @note See documentation of parallel::distributed::SolutionTransfer for
+     * matching code snippets for both transfer and serialization.
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2018
+     */
+    template <int dim, int spacedim = dim, typename VectorType = Vector<double>>
```
Comment (tamiko) [inline include/deal.II/distributed/cell_data_transfer.h]: You use a vector `typename VectorType` here. Can you comment in the class description how you assume this vector to be ordered (and what the underlying cell order is)? (For example, it would be equally possible to store cell-related data in a `std::map` with an active cell iterator as key.)

File: include/deal.II/distributed/cell_data_transfer.h [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+       */
+      void
+      prepare_for_coarsening_and_refinement(const VectorType &in);
+
+      /**
+       * Same as the function above, only for a list of vectors.
+       */
+      void
+      prepare_for_coarsening_and_refinement(
+        const std::vector<const VectorType *> &all_in);
```
Comment (tamiko) [inline include/deal.II/distributed/cell_data_transfer.h]: Can we find an alternative solution for this interface?

I am not a huge fan of `std::vector<const VectorType *> &` I have to say. Further, `prepare_for_coarsening_and_refinement` to register a (read-only) vector and `unpack` to read out the transferred values is not very intuitive.

What about the following high-level interface:
 - A single `register(VectorType &vec)`, and corresponding `deregister(VectorType &vec)` function that registers vectors that shall be maintained by the CellDataTransfer object.
 - When you register a number of vectors `vec1`, `vec2`, etc. the class automatically
   - makes a copy of the data before coarsening/refinement
   - resizes the referenced vectors after coarsening/refinement and writes out the new values immediately to `vec1`, `vec2`, and removes the temporary storage?

This would significantly simplify the interface :-)

Comment (tamiko) [reply include/deal.II/distributed/cell_data_transfer.h]: Later, if you want to register a container storing vectors you can simply do:
```
for(auto & it : container_with_vectors)
  cell_data_transfer.register(it);
```
(or use `std::for_each`). There is no need to add inflexible wrappers such as `std::vector<VectorType *>` in the calling interface.

File: include/deal.II/distributed/cell_data_transfer.h [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+     * //[create coarse mesh...]
+     * triangulation.load(filename);
+     *
+     * parallel::distributed::CellDataTransfer<dim, spacedim, Vector<double>>
+     *   cell_data_trans(triangulation);
+     * Vector<double> transferred_data (triangulation.n_active_cells());
+     * cell_data_trans.deserialize(transferred_data);
+     * @endcode
+     *
+     * @note If you use more than one object to attach data to a
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: say "more than one CellDataTransfer object" to make it clear what kind of object you're talking about

Comment (bangerth) [reply include/deal.II/distributed/cell_data_transfer.h]: or do you want to be more general here?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+     * parallel::distributed::Triangulation at the same time (e.g. a
+     * parallel::distributed::SolutionTransfer object) with the aim of
+     * serialization, the calls to
+     * parallel::distributed::CellDataTransfer::prepare_serialization()
+     * and parallel::distributed::SolutionTransfer::prepare_serialization(),
+     * as well as parallel::distributed::CellDataTransfer::deserialize() and
+     * parallel::distributed::SolutionTransfer::deserialize(),
+     * need to be in the same order, respectively.
+     *
+     * @note See documentation of parallel::distributed::SolutionTransfer for
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]:
```suggestion
     * @note See the documentation of parallel::distributed::SolutionTransfer for
```

File: include/deal.II/distributed/cell_data_transfer.h (hunk 8 above)
Comment (bangerth) [reply include/deal.II/distributed/cell_data_transfer.h]: The order, as in many other places, is the order in which active cells are traversed.

File: include/deal.II/distributed/cell_data_transfer.h [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    {
+    public:
+      /**
+       * Whenever data shall be transferred across coarsening, it is not a
+       * trivial choice how to process data of cells which will be coarsened. Or
+       * in other words, which data should be stored in the corresponding parent
+       * cell.
+       *
+       * In this namespace, we offer a few strategies that cope with this
+       * problem, that can be passed to the
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]:
```suggestion
       * problem, and that can be then passed to the
```

File: include/deal.II/distributed/cell_data_transfer.h [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+      struct CoarseningStrategies
+      {
+        /**
+         * Evaluate data from @p input_vector on children of @p parent.
+         * Check if data on all children match, and return that value.
+         */
+        static typename VectorType::value_type
+        check_equality(const typename parallel::distributed::
+                         Triangulation<dim, spacedim>::cell_iterator &parent,
+                       const VectorType *input_vector)
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: can you adjust the interface so that a reference instead of a pointer is passed here? same for the other functions of course

File: include/deal.II/distributed/cell_data_transfer.h [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+      };
+
+      /**
+       * Constructor.
+       *
+       * @param[in] triangulation The triangulation on which all operations will
+       *   happen. At the time when this constructor is called, the refinement
+       *   in question has not happened yet.
+       * @param[in] transfer_variable_size_data Specify whether size of data to
+       *   transfer varies from cell to cell.
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: can you elaborate here what that means? I mean, I'm passing a `std::vector<unsigned int>` in, that looks pretty constant to me... :-)

File: include/deal.II/distributed/cell_data_transfer.h [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+       * @param[in] coarsening_strategy Function deciding which data to store on
+       *   a cell, whose children will get coarsened into.
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: no comma in the English language

File: include/deal.II/distributed/cell_data_transfer.h [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+        const std::vector<const VectorType *> &all_in);
+
+      /**
+       * Prepare the serialization of the given vector. The serialization is
+       * done by Triangulation::save(). The given vector needs all information
+       * on the locally active DoFs (it must be ghosted). See documentation of
+       * this class for more information.
+       */
+      void
+      prepare_serialization(const VectorType &in);
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: this is the place where I would suggest to name the function `prepare_for_serialization()` (same below). Or have we made the same mistake in other places already?

Comment (bangerth) [reply include/deal.II/distributed/cell_data_transfer.h]: Different question: On the same object, can this function be called more than once with different vectors? I suspect the answer is "no", but it might be worth mentioning, with a hint towards the next function.

File: include/deal.II/distributed/cell_data_transfer.h [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+       * Same as the function above, only for a list of vectors.
+       */
+      void
+      prepare_for_coarsening_and_refinement(
+        const std::vector<const VectorType *> &all_in);
+
+      /**
+       * Prepare the serialization of the given vector. The serialization is
+       * done by Triangulation::save(). The given vector needs all information
+       * on the locally active DoFs (it must be ghosted). See documentation of
```
Comment (bangerth) [inline include/deal.II/distributed/cell_data_transfer.h]: the middle sentence here is not correct

File: include/deal.II/distributed/cell_data_transfer.h (hunk 9 above)
Comment (bangerth) [reply include/deal.II/distributed/cell_data_transfer.h]: @tamiko -- that's a design decision copied from the `SolutionTransfer` class. We made this mistake 20 years ago, and I'm in favor of using the same interface in all of these classes, rather than breaking it for one class but leaving the others as is.

Comment (tamiko) [reply include/deal.II/distributed/cell_data_transfer.h]: Well, then let's go that route. (And think about a better design later.)

File: include/deal.II/distributed/cell_data_transfer.h (hunk 8 above)
Comment (tamiko) [reply include/deal.II/distributed/cell_data_transfer.h]: I know, but it would be good to mention this here again.

File: include/deal.II/distributed/cell_data_transfer.h [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+     *
+     * For serialization, the following code snippet saves not only the
+     * triangulation itself, but also the cell by cell data attached:
+     * @code
+     * Vector<double> data_to_transfer (triangulation.n_active_cells());
+     * //[fill data_to_transfer with cell by cell values...]
+     *
+     * parallel::distributed::CellDataTransfer<dim, spacedim, Vector<double>>
+     *   cell_data_trans(triangulation);
+     * cell_data_trans.prepare_serialization(data_to_transfer);
```
[the remark this answers is not in the document]
Comment (marcfehling) [reply include/deal.II/distributed/cell_data_transfer.h]: I named it after the corresponding `p::d::SolutionTransfer::prepare_serialization()` function to maintain a similar interface. Shall we leave it without the `for`, or include it for both classes?

File: include/deal.II/distributed/cell_data_transfer.h (hunk 10 above)
Comment (marcfehling) [reply include/deal.II/distributed/cell_data_transfer.h]: Any object using the `register_data_attach` and `notify_ready_for_unpack` interface. I know that`SolutionTransfer` and `ParticleHandler` are using it. I can be more specific on that.

File: include/deal.II/distributed/cell_data_transfer.h (hunk 16 above)
Comment (marcfehling) [reply include/deal.II/distributed/cell_data_transfer.h]: We already named these function `prepare_serialization` in the `SolutionTransfer` class, so I adopted the interface.

You are right, the answer is 'no'. I updated the documentation: The function can be called once until transfer completed.

**Comment by marcfehling** [comment]
Took care of (most) suggestions. There are a few open questions for which I kindly ask your feedback. Ready for tests.

**Comment by marcfehling** [comment]
We need the `CoarseningStrategies` namespace also for a non-distributed variant of the `CellDataTransfer` class. I will move those functions somewhere else in a follow-up pull request which provides the non-distributed variant.

File: include/deal.II/distributed/cell_data_transfer.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/distributed/cell_data_transfer.h]: Let's call it `prepare_FOR_serialization`. Then we'll rename the corresponding function in `SolutionTransfer` accordingly. See #7566.

**Comment by marcfehling** [comment]
I've implemented everything I wanted in this pull request. Ready for tests/merge.

Future pull requests will:
- replace `p::d::ActiveFEIndicesTransfer` and let `hp::DoFHandler` transfer its `active_fe_indices` automagically ;)
- provide a non-distributed version of `CellDataTransfer`

File: include/deal.II/distributed/cell_data_transfer.h [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+     * Transfer data that is associated with each active cell (like error
+     * indicators) while refining and/or coarsening a distributed triangulation
+     * and handle the necessary communication.
+     *
+     * This class therefore does for cell-related information what
+     * parallel::distributed::SolutionTransfer does for the values of degrees of
+     * freedom defined on a parallel::distributed::Triangulation.
+     *
+     * This class has been designed to operate on any kind of datatype that is
+     * serializable. A non-distributed Vector container has to be provided,
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: Could you please add a test which transfers a pair if integers (something simple like 1 cell -> 4 cells)?
I see that you assume existence of `VectorType::value_type` so it should work nicely with `std::vector<std::pair<int,int>>`.

File: include/deal.II/distributed/cell_data_transfer.h [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+     *
+     * @ingroup distributed
+     * @author Marc Fehling, 2018
+     */
+    template <int dim, int spacedim = dim, typename VectorType = Vector<double>>
+    class CellDataTransfer
+    {
+    public:
+      /**
+       * Whenever data shall be transferred across coarsening, it is not a
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: maybe something like: 

When the date is transferred during coarsening, it is not trivial to decide how to process the data of child cells which will be coarsened.

?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+       * trivial choice how to process data of cells which will be coarsened. Or
+       * in other words, which data should be stored in the corresponding parent
+       * cell.
+       *
+       * In this namespace, we offer a few strategies that cope with this
+       * problem, of which one can be picked and passed to the
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: ... problem. Such strategies can be passed to the ... (?)

File: include/deal.II/distributed/cell_data_transfer.h [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+        }
+      };
+
+      /**
+       * Constructor.
+       *
+       * @param[in] triangulation The triangulation on which all operations will
+       *   happen. At the time when this constructor is called, the refinement
+       *   in question has not happened yet.
+       * @param[in] transfer_variable_size_data Specify whether your VectorType
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: could you please add a test which checks this functionality?

File: include/deal.II/distributed/cell_data_transfer.h [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+      /**
+       * Specifies if size of data to transfer varies from cell to cell.
+       */
+      const bool transfer_variable_size_data;
+
+      /**
+       * Function deciding which data to store from @p input_vector on a
+       * cell @p parent, whose children will get coarsened into.
+       */
+      std::function<typename VectorType::value_type(
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: that's also `const`

File: include/deal.II/distributed/cell_data_transfer.h [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+       *   container stores values that differ in size. A varying amount of data
+       *   may be packed per cell, if for example the underlying ValueType of
+       *   the VectorType container is a container itself.
+       * @param[in] coarsening_strategy Function deciding which data to store on
+       *   a cell whose children will get coarsened into.
+       */
+      CellDataTransfer(const parallel::distributed::Triangulation<dim, spacedim>
+                         &        triangulation,
+                       const bool transfer_variable_size_data = false,
+                       std::function<typename VectorType::value_type(
```
Comment (davydden) [inline include/deal.II/distributed/cell_data_transfer.h]: `const`

File: include/deal.II/distributed/cell_data_transfer.h (hunk 22 above)
Comment (marcfehling) [reply include/deal.II/distributed/cell_data_transfer.h]: The `transfer_variable_size_data` parameter will be passed to the `p::d::Triangulation::register_data_attach()` function. We already have a lot of tests for this functionality, like `mpi/attach_data_02`, `mpi/solution_transfer_04`, and `mpi/p4est_save_06`.

I would refrain from adding another.

File: include/deal.II/distributed/cell_data_transfer.h (hunk 19 above)
Comment (marcfehling) [reply include/deal.II/distributed/cell_data_transfer.h]: This would be possible, but from my understanding, we would have to supply another explicit template specialization for it (as in `cell_data_transfer.inst.in`).

Do we want to do that? Or is there a way around it that I couldn't come up with?

Comment (davydden) [reply include/deal.II/distributed/cell_data_transfer.h]: > Do we want to do that?

no, we don't. We should move all the definitions to the `.templates.h`, instantiate for things we find most useful (like `Vector<double>`) but still let users include both headers (i.e. `cell_data_transfer.h` and `cell_data_transfer.templates.h`) to be used with custom data types and containers so long as they provide things like `value_type`.

p.s. @bangerth please correct me if I am wrong in the interpretation of the `.templates.h` and `.h` in deal.II.

File: source/distributed/cell_data_transfer.cc [hunk 25]
```diff
@@ -0,0 +1,339 @@
...
+                *it_output = coarsening_strategy(cell, *(*it_input));
+                break;
+
+              default:
+                Assert(false, ExcInternalError());
+                break;
+            }
+        }
+
+      // We don't have to pack the whole container if there is just one entry.
```
Comment (davydden) [inline source/distributed/cell_data_transfer.cc]: that comment does not seem to be related to the actual code below.

File: include/deal.II/distributed/cell_data_transfer.h (hunk 22 above)
Comment (davydden) [reply include/deal.II/distributed/cell_data_transfer.h]: true, but you also used in other places, like pack/unpack.

**Comment by tamiko** [comment]
@marcfehling Please ping when we should review and approve :-)

**Comment by marcfehling** [comment]
Moved definitions to a dedicated `templates.h` file. Added tests for serialization and the transfer of a `std::vector<int>` value type with variable sizes.

Ready for review and testing! Let me know once you are happy with it, then I'll squash all commits into one.
